README.md
# phone

[![Build Status](https://travis-ci.org/lykmapipo/phone.svg?branch=master)](https://travis-ci.org/lykmapipo/phone)
[![Dependencies Status](https://david-dm.org/lykmapipo/phone.svg)](https://david-dm.org/lykmapipo/phone)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/phone/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/phone?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/phone)](https://github.com/lykmapipo/phone/blob/master/LICENSE) 

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/@lykmapipo/phone)](https://www.npmjs.com/package/@lykmapipo/phone)

Helper utilities for parsing and validate phone numbers

## Requirements

- [NodeJS v12+](https://nodejs.org)
- [Npm v6+](https://www.npmjs.com/)

## Installation

```sh
npm install --save @lykmapipo/phone
```

## Usage

```js
import { parsePhoneNumber } from '@lykmapipo/phone';

const phoneNumber = parsePhoneNumber('+255715333777', 'TZ');

//=> result
{ 
  raw: '+255715333777',
  countryCode: 'TZ',
  callingCode: 255,
  extension: '',
  isValid: true,
  isPossible: true,
  isValidNumberForCountryCode: true,
  isFixedLine: false,
  isMobile: true,
  type: 'MOBILE',
  isFixedLineOrMobile: false,
  isTollFree: false,
  isPremiumRate: false,
  isSharedCost: false,
  isVoip: false,
  isPersonalNumber: false,
  isPager: false,
  isUan: false,
  isVoicemail: false,
  isUnknown: false,
  e164: '+255715333777',
  e164NoPlus: '255715333777',
  international: '+255 715 333 777',
  national: '0715 333 777',
  rfc3966: 'tel:+255-715-333-777'
}
```

## Environment
```js
DEFAULT_COUNTRY_CODES=
DEFAULT_COUNTRY_CODE=
```

## Test

- Clone this repository

- Install all dependencies

```sh
npm install
```

- Then run test

```sh
npm test
```

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## LICENSE

MIT License

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.7.15 (2021-08-09)

##### Chores

* **deps:**  force latest version & audit fix ([9561be64](https://github.com/lykmapipo/phone/commit/9561be64919d1526f8ee3bca9a8b648e01ed590b))

#### 0.7.14 (2021-06-01)

##### Chores

* **deps:**  force latest version & audit fix ([98f973ba](https://github.com/lykmapipo/phone/commit/98f973ba11d605af92f39e62bf61e779daee5b3d))

#### 0.7.13 (2021-03-21)

#### 0.7.12 (2021-02-09)

#### 0.7.11 (2021-01-04)

##### Chores

* **deps:**  force latest version & audit fix ([be86f78e](https://github.com/lykmapipo/phone/commit/be86f78e37889641aadbd283b65bebe01e5bbce0))

#### 0.7.10 (2020-12-25)

##### Chores

* **deps:**  force latest version & audit fix ([564874a9](https://github.com/lykmapipo/phone/commit/564874a9b9eeda3866069533cac92a9c52ea8639))

#### 0.7.9 (2020-12-16)

##### Chores

* **deps:**  force latest version & audit fix ([dd1d02d8](https://github.com/lykmapipo/phone/commit/dd1d02d8a7ced65034a513f142bd9ead246ac844))

#### 0.7.8 (2020-12-05)

##### Chores

* **deps:**  force latest version & audit fix ([8896555c](https://github.com/lykmapipo/phone/commit/8896555cac6a9bc84ccc08c0070902056c5b3bb9))

#### 0.7.7 (2020-11-13)

##### Chores

* **deps:**  force latest version & audit fix ([0044b929](https://github.com/lykmapipo/phone/commit/0044b929094916858dd4f94efaffaf205b2a4066))

#### 0.7.6 (2020-10-05)

##### Chores

* **deps:**  force latest version & audit fix ([ffb24ce9](https://github.com/lykmapipo/phone/commit/ffb24ce9cb54f93914f6429836a0f94b97684f00))

#### 0.7.5 (2020-09-23)

#### 0.7.4 (2020-09-10)

##### Chores

* **deps:**  force latest version & audit fix ([51cdfce6](https://github.com/lykmapipo/phone/commit/51cdfce6a67e09dce2b3378e74cdc50a9659f697))

#### 0.7.3 (2020-09-02)

##### Chores

* **deps:**  force latest version & audit fix ([3aff29a5](https://github.com/lykmapipo/phone/commit/3aff29a5e17b8196f6078489bd62200952308367))

#### 0.7.2 (2020-08-19)

##### Chores

* **deps:**  force latest version & audit fix ([b3ce50af](https://github.com/lykmapipo/phone/commit/b3ce50af6cf85f93feeadb8e11ec3df72850bc8f))

#### 0.7.1 (2020-08-04)

##### Chores

* **deps:**  force latest version & audit fix ([6a85da48](https://github.com/lykmapipo/phone/commit/6a85da4810b67addde6b2a0352239d58865c1763))

#### 0.7.0 (2020-08-01)

##### Chores

* **deps:**
  *  force latest version & audit fix ([42966372](https://github.com/lykmapipo/phone/commit/429663724d3c391604d03671f7f21b9945efbec3))
  *  fix prototype pollution vulnerability in dot-prop ([1b141125](https://github.com/lykmapipo/phone/commit/1b14112588cc9bb6a0817f6f459aac250b8cbca3))
  *  force latest version & audit fix ([991b8c8a](https://github.com/lykmapipo/phone/commit/991b8c8a34e410ad48354b46ec6c6cc4cbb7c0e4))
  *  force latest version & audit fix ([35c5f245](https://github.com/lykmapipo/phone/commit/35c5f245c3da035615b66b348549852469a03cd3))
* **config:**  correct build external ([2be499fe](https://github.com/lykmapipo/phone/commit/2be499fec1fa3727cdde6267a2a09196267d5fae))

##### New Features

* **utils:**
  *  add type validity check ([104aac83](https://github.com/lykmapipo/phone/commit/104aac8323ecbc5a1012bfa332919cef87019652))
  *  add phone number formatters ([30648fdc](https://github.com/lykmapipo/phone/commit/30648fdc8b1aea43bc8383d032b856c387ef6eb2))
  *  add parse raw phone number ([23ca4d9d](https://github.com/lykmapipo/phone/commit/23ca4d9dddf2652954597d7a0abd4149101a030a))

##### Refactors

* **utils:**
  *  improve parsing by country code ([10c779b0](https://github.com/lykmapipo/phone/commit/10c779b081df2bc9c86cbb40686e97b6828dd2b6))
  *  rename format to applyFormats ([8e42c1b7](https://github.com/lykmapipo/phone/commit/8e42c1b72573ae26f0340e11e9a85c1cf1292785))
  *  rename checkValidity to checkTypes ([4d2d1777](https://github.com/lykmapipo/phone/commit/4d2d17775c33d2cdad7198aa819c66253d1e12fa))
*  slice parsePhoneNumber to utils ([ed184d87](https://github.com/lykmapipo/phone/commit/ed184d87230c5a8f5d432837f27479f764e21aaa))
*  extract constants from utils ([0a94d27b](https://github.com/lykmapipo/phone/commit/0a94d27bfede15c582325c55fae07bdfe7e25b61))
*  remove default exports & add utils ([64311a09](https://github.com/lykmapipo/phone/commit/64311a09c444583a5c08a500c2c6b305d57ee329))
*  use lodash specific imports ([ae0f2e48](https://github.com/lykmapipo/phone/commit/ae0f2e488b5e71d72308e7cde2d49edde7f05b3c))
*  remove underscore dangles ([e37cb02d](https://github.com/lykmapipo/phone/commit/e37cb02dbed1f0e13de2c5bb43ad3fd85a8c059f))
*  re-write use es6 ([3115a878](https://github.com/lykmapipo/phone/commit/3115a8787f5f9c9bb04461da293a6117f2d9a656))

##### Code Style Changes

*  improve jsdocs ([ea5ce8de](https://github.com/lykmapipo/phone/commit/ea5ce8de7a0753664ab2896f15db17ab5c27c500))
*  clear unused demarcations ([d9589a71](https://github.com/lykmapipo/phone/commit/d9589a7184335d5a0f2b51f4bf93a880316575e6))
*  add TYPES & FORMATS jsdocs ([e9997483](https://github.com/lykmapipo/phone/commit/e9997483a9a6124c358075391eeddbf4583dfd05))
*  improve jsdocs ([ea6e2097](https://github.com/lykmapipo/phone/commit/ea6e20973cfdc8ee00a6be29deb9162b01159f4d))

#### 0.6.11 (2020-07-20)

##### Chores


test/utils.spec.js
import { expect } from '@lykmapipo/test-helpers';
import { FORMAT_E164 } from '../src/constants';
import {
  phoneNumberUtil,
  parseRawPhoneNumber,
  formatPhoneNumber,
  applyFormats,
  checkTypes,
  parsePhoneNumberByCountryCode,
} from '../src/utils';

describe('phone utils', () => {
  it('should exports phone number util instance', () => {
    expect(phoneNumberUtil).to.exist;
    expect(phoneNumberUtil.parseAndKeepRawInput).to.exist.and.be.a('function');
  });

  it('should parse raw phone number', () => {
    expect(parseRawPhoneNumber).to.exist.and.be.a('function');
    expect(parseRawPhoneNumber('+255715333777')).to.exist;
    expect(parseRawPhoneNumber('+255715333777', 'TZ')).to.exist;
    expect(parseRawPhoneNumber('255715333777', 'TZ')).to.exist;
    expect(parseRawPhoneNumber('0715333777')).to.exist;
    expect(parseRawPhoneNumber('0715333777', 'TZ')).to.exist;
    expect(parseRawPhoneNumber('0715333777', 'TTT')).to.be.undefined;
    expect(parseRawPhoneNumber(null)).to.be.undefined;
    expect(parseRawPhoneNumber(undefined)).to.be.undefined;
    expect(parseRawPhoneNumber('')).to.be.undefined;
    expect(parseRawPhoneNumber(' ')).to.be.undefined;
  });

  it('should format raw phone number', () => {
    expect(formatPhoneNumber).to.exist.and.be.a('function');

    expect(
      formatPhoneNumber(parseRawPhoneNumber('+255715333777'), FORMAT_E164)
    ).to.be.eql({ e164: '+255715333777' });

    expect(
      formatPhoneNumber(parseRawPhoneNumber('+255715333777', 'TZ'), FORMAT_E164)
    ).to.be.eql({ e164: '+255715333777' });

    expect(
      formatPhoneNumber(parseRawPhoneNumber('255715333777', 'TZ'), FORMAT_E164)
    ).to.be.eql({ e164: '+255715333777' });

    expect(
      formatPhoneNumber(parseRawPhoneNumber('0715333777'), FORMAT_E164)
    ).to.be.eql({ e164: '+255715333777' });

    expect(
      formatPhoneNumber(parseRawPhoneNumber('0715333777', 'TZ'), FORMAT_E164)
    ).to.be.eql({ e164: '+255715333777' });

    expect(
      formatPhoneNumber(parseRawPhoneNumber('0715333777', 'TTT'), FORMAT_E164)
    ).to.be.eql({});

    expect(formatPhoneNumber(parseRawPhoneNumber(null), FORMAT_E164)).to.be.eql(
      {}
    );

    expect(
      formatPhoneNumber(parseRawPhoneNumber(undefined), FORMAT_E164)
    ).to.be.eql({});
  });

  it('should format phone number with all available formats', () => {
    expect(applyFormats).to.exist.and.be.a('function');

    expect(applyFormats(parseRawPhoneNumber('+255715333777'))).to.be.eql({
      e164: '+255715333777',
      international: '+255 715 333 777',
      national: '0715 333 777',
      rfc3966: 'tel:+255-715-333-777',
    });
  });

  it('should check phone number types validity', () => {
    expect(checkTypes).to.be.a('function');
    expect(checkTypes(parseRawPhoneNumber('+255715333777'))).to.be.eql({
      isFixedLine: false,
      isFixedLineOrMobile: false,
      isMobile: true,
      isPager: false,
      isPersonalNumber: false,
      isPremiumRate: false,
      isSharedCost: false,
      isTollFree: false,
      isUan: false,
      isUnknown: false,
      isVoicemail: false,
      isVoip: false,
      type: 'MOBILE',
    });
  });

  it('should parse phone number', () => {
    const rawPhoneNumber = '+255715333777';
    const countryCode = 'TZ';
    const phoneNumber = parsePhoneNumberByCountryCode(
      rawPhoneNumber,
      countryCode
    );
    expect(phoneNumber).to.exist;
    expect(phoneNumber.raw).to.exist;
    expect(phoneNumber.countryCode).to.exist;
    expect(phoneNumber.callingCode).to.exist;
    expect(phoneNumber.extension).to.exist;
    expect(phoneNumber.isValid).to.exist;
    expect(phoneNumber.isPossible).to.exist;
    expect(phoneNumber.isValidForCountryCode).to.exist;
    expect(phoneNumber.isFixedLine).to.exist;
    expect(phoneNumber.isMobile).to.exist;
    expect(phoneNumber.type).to.exist;
    expect(phoneNumber.isFixedLineOrMobile).to.exist;
    expect(phoneNumber.isTollFree).to.exist;
    expect(phoneNumber.isPremiumRate).to.exist;
    expect(phoneNumber.isSharedCost).to.exist;
    expect(phoneNumber.isVoip).to.exist;
    expect(phoneNumber.isPersonalNumber).to.exist;
    expect(phoneNumber.isPager).to.exist;
    expect(phoneNumber.isUan).to.exist;
    expect(phoneNumber.isVoicemail).to.exist;
    expect(phoneNumber.isUnknown).to.exist;
    expect(phoneNumber.e164).to.exist;
    expect(phoneNumber.international).to.exist;
    expect(phoneNumber.national).to.exist;
    expect(phoneNumber.rfc3966).to.exist;
  });

  it('should parse phone number', () => {
    const rawPhoneNumber = null;
    const countryCode = 'TZ';
    const phoneNumber = parsePhoneNumberByCountryCode(
      rawPhoneNumber,
      countryCode
    );
    expect(phoneNumber).to.be.undefined;
  });
});


test/bootstrap.spec.js
process.env.DEFAULT_COUNTRY_CODE = 'TZ';
process.env.DEFAULT_COUNTRY_CODES = 'TZ,US';


lib/index.js
'use strict';

const clone = require('lodash/clone');
const find = require('lodash/find');
const map = require('lodash/map');
const toUpper = require('lodash/toUpper');
const common = require('@lykmapipo/common');
const env = require('@lykmapipo/env');
const camelCase = require('lodash/camelCase');
const forEach = require('lodash/forEach');
const merge = require('lodash/merge');
const toLower = require('lodash/toLower');
const googleLibphonenumber = require('google-libphonenumber');
const keys = require('lodash/keys');

/**
 * @constant
 * @name TYPES
 * @description Allowed phone number types
 * @type {Array}
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * import { TYPES, TYPE_MOBILE } from '@lykmapipo/phone';
 *
 * TYPES //=> [ 'MOBILE', ... ];
 * TYPE_MOBILE //=> 'MOBILE';
 */
const TYPES = common.sortedUniq(keys(googleLibphonenumber.PhoneNumberType));

// types
const TYPE_FIXED_LINE = 'FIXED_LINE';
const TYPE_FIXED_LINE_OR_MOBILE = 'FIXED_LINE_OR_MOBILE';
const TYPE_MOBILE = 'MOBILE';
const TYPE_PAGER = 'PAGER';
const TYPE_PERSONAL_NUMBER = 'PERSONAL_NUMBER';
const TYPE_PREMIUM_RATE = 'PREMIUM_RATE';
const TYPE_SHARED_COST = 'SHARED_COST';
const TYPE_TOLL_FREE = 'TOLL_FREE';
const TYPE_UAN = 'UAN';
const TYPE_UNKNOWN = 'UNKNOWN';
const TYPE_VOICEMAIL = 'VOICEMAIL';
const TYPE_VOIP = 'VOIP';

/**
 * @constant
 * @name FORMATS
 * @description Allowed phone number formats
 * @type {Array}
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * import { FORMATS, FORMAT_E164 } from '@lykmapipo/phone';
 *
 * FORMATS //=> [ 'E164', ... ];
 * FORMAT_E164 //=> 'E164';
 */
const FORMATS = common.sortedUniq(keys(googleLibphonenumber.PhoneNumberFormat));

// formats
const FORMAT_E164 = 'E164';
const FORMAT_INTERNATIONAL = 'INTERNATIONAL';
const FORMAT_NATIONAL = 'NATIONAL';
const FORMAT_RFC3966 = 'RFC3966';

/**
 * @name phoneNumberUtil
 * @function phoneNumberUtil
 * @description Internal instance of `PhoneNumberUtil`
 * @returns {object} Validity of `PhoneNumberUtil`
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.5.0
 * @version 0.1.0
 * @private
 * @example
 *
 * phoneNumberUtil.parseAndKeepRawInput('0715333777', 'TZ');
 *
 */
const phoneNumberUtil = googleLibphonenumber.PhoneNumberUtil.getInstance();

/**
 * @name parseRawPhoneNumber
 * @function parseRawPhoneNumber
 * @description Attempt to parse given phone number
 * @param {string} phoneNumber Number that we are attempting to parse
 * @param {string} [countryCode] 2 or 3 letter ISO country code. Default to
 * `process.env.DEFAULT_COUNTRY_CODE` or `os country code`.
 * @returns {object} Validity of `PhoneNumberUtil`
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.5.0
 * @version 0.1.0
 * @private
 * @example
 *
 * parseRawPhoneNumber('+255715333777');
 * parseRawPhoneNumber('0715333777', 'TZ');
 *
 */
const parseRawPhoneNumber = (
  phoneNumber,
  countryCode = env.getCountryCode()
) => {
  try {
    // try parse phone number
    const rawPhoneNumber = clone(phoneNumber);
    const rawCountryCode = clone(countryCode);
    const parsedPhoneNumber = phoneNumberUtil.parseAndKeepRawInput(
      rawPhoneNumber,
      rawCountryCode
    );
    return parsedPhoneNumber;
  } catch (e) {
    // fail to parse phone number
    return undefined;
  }
};

/**
 * @name formatPhoneNumber
 * @function formatPhoneNumber
 * @description Format phone number use given format
 * @param {object} phoneNumber Valid instance of parsed phone number
 * @param {string} [format] Valid phone number format
 * @returns {object} Formatted phone number
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.5.0
 * @version 0.1.0
 * @private
 * @example
 *
 * const phoneNumber = parseRawPhoneNumber('0715333777', 'TZ');
 * formatPhoneNumber(phoneNumber, FORMAT_E164);
 *
 * //=> result
 * {
 *   e164: '+255715333777'
 * }
 *
 */
const formatPhoneNumber = (phoneNumber, format) => {
  try {
    // try format parsed phone number
    const phoneNumberFormat = toLower(format);
    const formattedPhoneNumber = phoneNumberUtil.format(
      phoneNumber,
      googleLibphonenumber.PhoneNumberFormat[format]
    );
    return { [phoneNumberFormat]: formattedPhoneNumber };
  } catch (e) {
    // fail to format phone number
    return {};
  }
};

/**
 * @name applyFormats
 * @function applyFormats
 * @description Format phone number using available phone number formats
 * @param {object} phoneNumber Instance of parsed phone number
 * @returns {object} Formatted phone number(s)
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.2.0
 * @private
 * @example
 *
 * const phoneNumber = parseRawPhoneNumber('0715333777', 'TZ');
 * applyFormats(phoneNumber);
 *
 * //=> result
 * {
 *  e164: '+255715333777',
 *  international: '+255 715 333 777',
 *  national: '0715 333 777',
 *  rfc3966: 'tel:+255-715-333-777'
 * }
 *
 */
const applyFormats = (phoneNumber) => {
  // initialize formats
  let formats = {};

  // format phone number per each available format
  forEach(FORMATS, (phoneNumberFormat) => {
    const formattedPhoneNumber = formatPhoneNumber(
      phoneNumber,
      phoneNumberFormat
    );
    formats = common.mergeObjects(formats, formattedPhoneNumber);
  });

  // return formatted phone number
  return formats;
};

/**
 * @name checkTypes
 * @function checkTypes
 * @description Derive phone number type validity
 * @param {object} phoneNumber Instance of parsed phone number
 * @returns {object} Validities of a phone number
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.5.0
 * @version 0.1.0
 * @private
 * @example
 *
 * const phoneNumber = parseRawPhoneNumber('0715333777', 'TZ');
 * checkTypes(phoneNumber);
 *
 * //=> result
 * {
 *  isFixedLine: false,
 *  isMobile: true,
 *  isFixedLineOrMobile: false,
 *  isTollFree: false,
 *  isPremiumRate: false,
 *  isSharedCost: false,
 *  isVoip: false,
 *  isPersonalNumber: false,
 *  isPager: false,
 *  isUan: false,
 *  isVoicemail: false,
 *  isUnknown: false,
 *  type: 'MOBILE'
 * }
 *
 */
const checkTypes = (phoneNumber) => {
  // initialize types validity
  const types = {};

  // obtain parsed phone number type
  const phoneNumberType = common.tryCatch(
    () => phoneNumberUtil.getNumberType(phoneNumber),
    -1
  );

  // type checker
  const checkType = (typeIndex, typeName) => {
    // derive type name and check phone number type
    const numberTypeName = camelCase(`is${typeName}`);
    const numberTypeIs = phoneNumberType === typeIndex;

    // set type is flag
    types[numberTypeName] = numberTypeIs;

    // set type name string
    if (numberTypeIs) {
      types.type = typeName;
    }
  };

  // check phone number type validity
  forEach(googleLibphonenumber.PhoneNumberType, checkType);

  // return types validity
  return types;
};

/**
 * @name parsePhoneNumberByCountryCode
 * @function parsePhoneNumberByCountryCode
 * @description Parse provided phone number to obtain its information
 * @param {string} phoneNumber Valid phone number
 * @param {...string} [countryCode] Valid country code(s) for validation. If
 * not provided `process.env.DEFAULT_COUNTRY_CODE` or `os country code` will
 * be used as a default
 * @returns {object | undefined} Parsed phone number or undefined
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.5.0
 * @version 0.1.0
 * @private
 * @static
 * @example
 *
 * const phoneNumber = parsePhoneNumberByCountryCode('+255715333777');
 * const phoneNumber = parsePhoneNumberByCountryCode('+255715333777', 'TZ');
 *
 * //=> result
 * {
 *  raw: '+255715333777',
 *  countryCode: 'TZ',
 *  callingCode: 255,
 *  extension: '',
 *  isValid: true,
 *  isPossible: true,
 *  isValidForCountryCode: true,
 *  isFixedLine: false,
 *  isMobile: true,
 *  type: 'MOBILE',
 *  isFixedLineOrMobile: false,
 *  isTollFree: false,
 *  isPremiumRate: false,
 *  isSharedCost: false,
 *  isVoip: false,
 *  isPersonalNumber: false,
 *  isPager: false,
 *  isUan: false,
 *  isVoicemail: false,
 *  isUnknown: false,
 *  e164: '+255715333777',
 *  international: '+255 715 333 777',
 *  national: '0715 333 777',
 *  rfc3966: 'tel:+255-715-333-777'
 * }
 *
 */
const parsePhoneNumberByCountryCode = (phoneNumber, countryCode) => {
  // parse phone number
  try {
    const parsed = parseRawPhoneNumber(phoneNumber, countryCode);

    // prepare parse phone number result
    let phone = {};

    // set raw phone number
    phone.raw = phoneNumber;

    // set phone country code
    phone.countryCode =
      phoneNumberUtil.getRegionCodeForNumber(parsed) || countryCode;

    // set phone country calling code
    phone.callingCode = parsed.getCountryCodeOrDefault();

    // set phone number extension
    phone.extension = parsed.getExtensionOrDefault();

    // set phone number valid flag
    phone.isValid = phoneNumberUtil.isValidNumber(parsed);

    // set possible flag
    phone.isPossible = phoneNumberUtil.isPossibleNumber(parsed);

    // set is valid for country(or region) code
    phone.isValidForCountryCode = phoneNumberUtil.isValidNumberForRegion(
      parsed,
      phone.countryCode
    );

    // set phone number type flags
    phone = merge({}, phone, checkTypes(parsed));

    // format phone number in accepted formats
    phone = merge({}, phone, applyFormats(parsed));

    // add e164 format with no plus
    phone.e164NoPlus = phone.e164.replace(/\+/g, '');

    // return parsed phone number
    return phone;
  } catch (e) {
    // fail to parse phone number
    return undefined;
  }
};

/**
 * @name parsePhoneNumber
 * @function parsePhoneNumber
 * @description Parse provided phone number to obtain its information
 * @param {string} phoneNumber Valid phone number
 * @param {...string} [countryCode] Valid country code(s) for validation. If
 * not provided `process.env.DEFAULT_COUNTRY_CODE` or `os country code` will
 * be used as a default
 * @returns {object | undefined} Parsed phone number or undefined
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @public
 * @static
 * @example
 *
 * import { parsePhoneNumber } from '@lykmapipo/phone';
 *
 * const phoneNumber = parsePhoneNumber('+255715333777');
 * const phoneNumber = parsePhoneNumber('+255715333777', 'TZ');
 *
 * //=> result
 * {
 *  raw: '+255715333777',
 *  countryCode: 'TZ',
 *  callingCode: 255,
 *  extension: '',
 *  isValid: true,
 *  isPossible: true,
 *  isValidForCountryCode: true,
 *  isFixedLine: false,
 *  isMobile: true,
 *  type: 'MOBILE',
 *  isFixedLineOrMobile: false,
 *  isTollFree: false,
 *  isPremiumRate: false,
 *  isSharedCost: false,
 *  isVoip: false,
 *  isPersonalNumber: false,
 *  isPager: false,
 *  isUan: false,
 *  isVoicemail: false,
 *  isUnknown: false,
 *  e164: '+255715333777',
 *  international: '+255 715 333 777',
 *  national: '0715 333 777',
 *  rfc3966: 'tel:+255-715-333-777'
 * }
 *
 */
const parsePhoneNumber = (phoneNumber, ...countryCode) => {
  try {
    // ensure raw phone number
    const raw = clone(phoneNumber);

    // collect country codes
    let countryCodes = env.getStringSet('DEFAULT_COUNTRY_CODES', env.getCountryCode());
    countryCodes = common.uniq([...countryCode, ...countryCodes]);
    countryCodes = common.uniq(map(countryCodes, toUpper));

    // test parsing for provided country codes
    const phones = map(countryCodes, (givenCountryCode) =>
      parsePhoneNumberByCountryCode(raw, givenCountryCode)
    );

    // return valid parsed or any
    const phone = find(phones, { isValid: true });
    return phone;
  } catch (error) {
    // fail to parse phone number
    return undefined;
  }
};

/**
 * @name toE164
 * @function toE164
 * @description Format provided mobile phone number to E.164 format
 * @param {string} phoneNumber a mobile phone number to be formatted
 * @param {string} [countryCode] 2 or 3 letter ISO country code
 * @returns {string} E.164 formatted phone number without leading plus sign
 * @see {@link https://en.wikipedia.org/wiki/E.164|e.164}
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @public
 * @static
 * @example
 *
 * import { toE164 } from '@lykmapipo/phone';
 *
 * const phoneNumber = toE164('0715333777');
 * const phoneNumber = toE164('0715333777', 'TZ');
 *
 * //=> result
 * 255715333777
 */
const toE164 = (phoneNumber, countryCode) => {
  // try convert give phone number to e.164
  try {
    const parsedNumber = parsePhoneNumber(phoneNumber, countryCode);
    if (parsedNumber && parsedNumber.e164NoPlus) {
      return parsedNumber.e164NoPlus;
    }
    return phoneNumber;
  } catch (error) {
    // fail to convert, return original format
    return phoneNumber;
  }
};

exports.FORMATS = FORMATS;
exports.FORMAT_E164 = FORMAT_E164;
exports.FORMAT_INTERNATIONAL = FORMAT_INTERNATIONAL;
exports.FORMAT_NATIONAL = FORMAT_NATIONAL;
exports.FORMAT_RFC3966 = FORMAT_RFC3966;
exports.TYPES = TYPES;
exports.TYPE_FIXED_LINE = TYPE_FIXED_LINE;
exports.TYPE_FIXED_LINE_OR_MOBILE = TYPE_FIXED_LINE_OR_MOBILE;
exports.TYPE_MOBILE = TYPE_MOBILE;
exports.TYPE_PAGER = TYPE_PAGER;
exports.TYPE_PERSONAL_NUMBER = TYPE_PERSONAL_NUMBER;
exports.TYPE_PREMIUM_RATE = TYPE_PREMIUM_RATE;
exports.TYPE_SHARED_COST = TYPE_SHARED_COST;
exports.TYPE_TOLL_FREE = TYPE_TOLL_FREE;
exports.TYPE_UAN = TYPE_UAN;
exports.TYPE_UNKNOWN = TYPE_UNKNOWN;
exports.TYPE_VOICEMAIL = TYPE_VOICEMAIL;
exports.TYPE_VOIP = TYPE_VOIP;
exports.parsePhoneNumber = parsePhoneNumber;
exports.toE164 = toE164;
